### PR TITLE
workerutil: experimenting with batch dequeue to reduce queue overhead

### PR DESCRIPTION
**Pull request by coury-clark**
### This PR is not meant to be merged, just looking for feedback here

So, what is this?

While debugging some slowness in the code insights worker, I noticed a few things:
1. The dequeue mechanism can be a bottleneck for short-lived, high concurrency `handlers` (dequeue operations can have non-trivial overhead)
2. Other queuing systems (kafka, sqs, etc) allow consumers to poll a batch of records with the semantics that records not successfully marked as completed in some time interval may be deqeueued again.

Now, I'm not interested in rebuilding Kafka here but it would be useful for code insights to have some batching. We are currently spending roughly 400ms on each dequeue operation (this could probably be a bit better, but regardless) and only 100-200ms average on the actual handler invocation. We have millions of invocations to process, so if we could batch something like 20-50 records we would nearly eliminate the queue overhead and save a *lot* of time.

Here are some of the implementation problems I'm looking for feedback on:
1. The [`RecordScanFn`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@workerutil/batch-dequeue/-/blob/internal/workerutil/dbworker/store/store.go?utm_product_name=GoLand&utm_product_version=GoLand&utm_source=JetBrains-v1.2.1#L211:18) can just brick the entire set of records currently because it eats the entire `rows` and gives back just one `Record`. I could update this to return a `[]workerutil.Record`, or I could add a `BatchRecordScanFn` with different semantics. Ideally, the operations on a single `Record` would be in a different store than operations to fetch a `Record` but I think that would be much more work.
2. Should the [`Store`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@workerutil/batch-dequeue/-/blob/internal/workerutil/dbworker/store/store.go?utm_product_name=GoLand&utm_product_version=GoLand&utm_source=JetBrains-v1.2.1#L87:11) interface have both a `Dequeue` and `BatchDequeue` operation? I'm leaning towards yes, to reduce the random impact of changing the `Dequeue` semantics. For example, in this PR there is an executor that uses the same `Store` interface that would have to be updated.
3. What to attach to the trace? Currently, it associates the single record `id` on dequeue. Is this even relevant?
4. Should batching respect the [`NumTotalJobs`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@workerutil/batch-dequeue/-/blob/internal/workerutil/dbworker/store/store.go?utm_product_name=GoLand&utm_product_version=GoLand&utm_source=JetBrains-v1.2.1#L87:11) exactly, or execute within a single `BatchSize`?

So basically this PR is a rough sketch implementation of what this could look like. It doesn't work right because the mapping function eats all of the records, but it should serve to illustrate the idea.

Looking for feedback, thanks!

### CODE REVIEW

File: internal/workerutil/worker.go [hunk 1]
```diff
@@ -95,6 +99,13 @@ func newWorker(ctx context.Context, store Store, handler Handler, options Worker
 		handlerSemaphore <- struct{}{}
 	}
 
+	cname := "src_" + options.Name + "worker_loop_total"
```
Comment (coury-clark) [inline internal/workerutil/worker.go:102]: This can be ignored, it's just for some testing

**Comment by efritz** [comment]
@coury-clark Ok I love this effort! I want to discuss a few points as there are some design decisions here. I'll base my opinions off of what I've had the most recent experience with in this area.

1. RabbitMQ has the concept of _prefetcing_, which will cache a message in the client before invoking a handler. We could do something like this, but we would end up still dequeueing and scanning reach record individually, just at a different (possibly slightly faster) rate than the handler is keeping up.
2. Amazon SQS does everything in batch, which seems a bit like what you've done here (and it's growing on me). As I understand it, SQS will give you zero or more messages up to `MaxNumberOfMessages` specified in the request, and you can change the visibility or delete the messages after processing all of them at once. It feels like this option is more in-line with the kind of work that is being done in the insights job.

Does this track so far? Are there any other options we can throw into the ring to consider for increasing throughput? Any cool Kafka achievements in this area?

**Comment by coury-clark** [comment]
> @coury-clark Ok I love this effort! I want to discuss a few points as there are some design decisions here. I'll base my opinions off of what I've had the most recent experience with in this area.
> 
>     1. RabbitMQ has the concept of _prefetcing_, which will cache a message in the client before invoking a handler. We could do something like this, but we would end up still dequeueing and scanning reach record individually, just at a different (possibly slightly faster) rate than the handler is keeping up.
> 
>     2. Amazon SQS does everything in batch, which seems a bit like what you've done here (and it's growing on me). As I understand it, SQS will give you zero or more messages up to `MaxNumberOfMessages` specified in the request, and you can change the visibility or delete the messages after processing all of them at once. It feels like this option is more in-line with the kind of work that is being done in the insights job.
> 
> 
> Does this track so far? Are there any other options we can throw into the ring to consider for increasing throughput? Any cool Kafka achievements in this area?

At a high-level Kafka lets a consumer lease records on a partition up to a specified time (that the consumer provides). It's the job of the consumer to "commit" the record (which moves the offset for that specific consumer group on that partition to the committed record). The advantage of the Kafka approach is that you can poll for a bunch of records (default 500 in many libraries) and process them at your own pace, as long as you obey the lease agreement. If you don't obey the lease agreement, the partitions may be given to a different consumer to read the same records. It ends up being substantially lower overhead on the queue (which is great for the Kafka use case).

The downside is that it's really complicated for everyone involved. There are a ton of consumer configuration options that can end up being a foot-gun if you don't really know Kafka, and there are a lot of error states that you can end up in even with a well-tuned cluster. I'd like to avoid the complexity of building a queue and have a simple scaffolding that lets us safely batch records with at-least once semantics.

I am leaning towards the SQS model, where we will poll for a batch of records that are owned by the `Worker` that acquired the lock. The records will have to be explicitly deleted (marked `complete` in our case) from the queue in order to avoid getting polled again. If a record fails during processing, we have the option to kick it ouf of the queue with the `failed` state or we can just skip it and it will get picked up in a future batch.

**Comment by efritz** [comment]
While we have this discussion, this may decrease code insight's dequeue latency: https://github.com/sourcegraph/sourcegraph/pull/24299

**Comment by tsenart** [comment]
I don't know if we're prepared to experiment with something different than Postgres for queueing, but if we were, Redis is already an infrastructure dependency that we ship with Sourcegraph and its Streams feature seems quite similar to Kafka semantics: https://redis.io/topics/streams-intro

**Comment by mrnugget** [comment]
@coury-clark Since this PR is "not meant to be merged" and there has been a bit of feedback, I removed backend-devs from the requested reviewers so I can get this PR out of my GitHub Slack bot notifications. If there's more input needed, just ping/re-add :)

**Comment by coury-clark** [comment]
> @coury-clark Since this PR is "not meant to be merged" and there has been a bit of feedback, I removed backend-devs from the requested reviewers so I can get this PR out of my GitHub Slack bot notifications. If there's more input needed, just ping/re-add :)

Thanks! My priority got sidetracked a bit these past few weeks - I'll be better about being proactive about removing reviewers in the future if this happens again.

**Comment by coury-clark** [comment]
Closing until we can revisit this with some time.
